### PR TITLE
os/bluestore: refactor _do_write(); move initializaiton of csum out of loop

### PR DESCRIPTION
**Pull request by xiexingguo**

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -8050,6 +8050,62 @@ void BlueStore::_do_write_data(
...
+    // we need garbage collection of blobs.
+    if (offset > gc_start_offset) {
+      bufferlist head_bl;
+      size_t read_len = offset - gc_start_offset;
+      int r = _do_read(c.get(), o, gc_start_offset, read_len, head_bl, 0);
+      assert(r == (int)read_len);
+      if (g_conf->bluestore_gc_merge_data) {
+        head_bl.claim_append(bl);
+        bl.swap(head_bl);
+        offset = gc_start_offset;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: offset & length can be updated within this func hence must be passed via reference

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -7892,20 +7906,6 @@ int BlueStore::_do_alloc_write(
       dblob.extents.push_back(e);
     }
 
-    // checksum
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: compression_required_ratio calculation above can be moved out of the loop too.

**Comment by xiexingguo** [comment]
@liewegas Fixed.

**Comment by voidbag** [comment]
Build check succeeded after rebasing master

**Pull request merged**